File: 백준/Gold/16987. 계란으로 계란치기/계란으로 계란치기.py
n = int(input())
arr = []
for _ in range(n):
    s, w = map(int, input().split())
    arr.append([s,w])

    # 내구도는 상대 계란의 무게만큼 깎이게 된다. 
    # 내구도가 0 이하가 되는 순간 계란은 깨지게 된다. 
ans  = 0
def dfs(x, cnt):
    global ans
    # 끝까지 가도 정답 x
    if ans >= (cnt + (n-x)*2): 
        return
    if x == n:
        ans = max(ans, cnt)
        return
    else:
# 손에 든계란 깨지면 다음으로
        if arr[x][0] <= 0:
            dfs(x+1, cnt)
        else:        
            flag = False
            for i in range(n):
                if i == x or arr[i][0] <= 0: 
                    continue
                flag = True
                arr[x][0] -= arr[i][1]
                arr[i][0] -= arr[x][1]
   
                dfs(x+1, cnt+int(arr[x][0]<=0)+int(arr[i][0]<=0))
                # 원상복구
                arr[x][0] += arr[i][1]
                arr[i][0] += arr[x][1]
            # 
            if flag == False:
                dfs(x+1, cnt)


dfs(0, 0)
print(ans)

File: 백준/Gold/16987. 계란으로 계란치기/계란으로 계란치기.java

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {
  static int n;
  static List<Egg> eggs;
  static int ans = 0;
  static class Egg{
    public int weight;
    public int durable;

    public Egg(int weight, int durable) {
      this.weight = weight;
      this.durable = durable;
    }

    public void fight(Egg other){
      this.durable -= other.weight;
      other.durable -= this.weight;
    }

    public void restore(Egg other){
      this.durable += other.weight;
      other.durable += this.weight;
    }

  }

  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    n = sc.nextInt();
    eggs = new ArrayList<>(n);
    for (int i = 0; i < n; i++) {
      int durable = sc.nextInt();
      int weight = sc.nextInt();
      eggs.add(new Egg(weight, durable));
    }
    recur(0,0);
    System.out.println(ans);

  }

  static void recur(int now, int cnt){
    if (now == n){
      ans = Math.max(ans, cnt);
      return;
    }

    boolean flag = false;
    if(eggs.get(now).durable > 0){

      for (int i = 0; i < n; i++) {
        if ( i == now || eggs.get(i).durable <= 0) continue;
        flag = true;
        eggs.get(now).fight(eggs.get(i));
        recur(now+1, cnt + (eggs.get(now).durable <= 0 ? 1: 0) + (eggs.get(i).durable <= 0 ? 1: 0)); ;
        eggs.get(now).restore(eggs.get(i));
      }
      if (!flag) recur(now+1, cnt);

    } else {
      recur(now+1, cnt);
    }

  }
}
